medcat/utils/saving/serializer.py

```diff
@@ -0,0 +1,202 @@
+import logging
+from typing import cast, Dict, Optional, Union
+import dill
+import json
+
+from medcat.config import Config
+logger = logging.getLogger(__name__)
+
+
+__SPECIALITY_NAMES_CUI = set(["cui2names", "cui2snames", "cui2type_ids"])
+__SPECIALITY_NAMES_NAME = set(
+    ["name2cuis", "name2cuis2status", "name_isupper"])
+__SPECIALITY_NAMES_OTHER = set(["snames"])
+SPECIALITY_NAMES = __SPECIALITY_NAMES_CUI | __SPECIALITY_NAMES_NAME | __SPECIALITY_NAMES_OTHER
+
+
+class SetEncode(json.JSONEncoder):
+    """JSONEncoder (and decoder) for sets.
+
+    Generally, JSON doesn't support serializing of sets natively.
+    This encoder adds a set identifier to the data when being serialized
+    and provides a method to read said identifier upon decoding."""
+    SET_IDENTIFIER = '==SET=='
+
+    def default(self, obj):
+        if isinstance(obj, set):
+            return {SetEncode.SET_IDENTIFIER: list(obj)}
+        return json.JSONEncoder.default(self, obj)
+
+    @staticmethod
+    def set_decode(dct: dict) -> Union[dict, set]:
+        """Decode sets from input dicts.
+
+        Args:
+            dct (dict): The input dict
+
+        Returns:
+            Union[dict, set]: The original dict if this was not a serialized set, the set otherwise
+        """
+        if SetEncode.SET_IDENTIFIER in dct:
+            return set(dct[SetEncode.SET_IDENTIFIER])
+        return dct
+
+
+class JsonSetSerializer:
+    """JSON serializer with set comprehension.
+
+    This serializer allows serializing and deserializing sets through JSON"""
+
+    def __init__(self, folder: str, name: str) -> None:
+        self.name = name
+        self.file_name = os.path.join(folder, name)
+        if not self.file_name.endswith('.json'):
+            self.file_name = self.file_name + '.json'
+        if not os.path.exists(folder):
+            os.makedirs(folder)
+        elif not os.path.isdir(folder):
+            raise ValueError(f'Folder expected, got fille: {folder}')
+        if os.path.isdir(self.file_name):
+            raise ValueError(
+                f'Expected file, found folder: {self.file_name}')
+
+    def write(self, d: dict) -> None:
+        """Write the specified dictionary to the this serializer's file.
+
+        Args:
+            d (dict): The dict to write on file.
+        """
+        logger.info('Writing data for "%s" into "%s"',
+                    self.name, self.file_name)
+        with open(self.file_name, 'w') as f:
+            json.dump(d, f, cls=SetEncode)
+
+    def read(self) -> dict:
+        """Read the json file specified by this serializer.
+
+        Returns:
+            dict: The dict represented by this json file.
+        """
+        logger.info('Reading data for %s from %s', self.name, self.file_name)
+        with open(self.file_name, 'r') as f:
+            data = json.load(f, object_hook=SetEncode.set_decode)
+        return data
+
+
+class CDBSerializer:
+    """A (potentially) semi-JSON based serializer for CDB.
```

Comment:
I kind of feel like one serialisation format should be used or another. If 'json' is selected everything should use JSON. 

Also in a future update can you also serialize the addl_info Dict?

Reply:
Good point. The current format is more like 'cdbpartjson' rather than 'json'.
Added two subtasks to the same ClickUp task regarding these two.